### PR TITLE
SPGISpeech recipe

### PR DESCRIPTION
**Pull request by desh2608**
This is a WIP PR for [SPGISpeech](https://arxiv.org/abs/2104.02014). I am opening this here early to discuss some issues I have been facing training a transducer model.

I first trained a conformer CTC model on the data (**without** speed perturbation) on 4 GPUs for 20 epochs, and the training curve looked reasonable: [tensorboard](https://tensorboard.dev/experiment/jduxUZlHRDa4OQRPSAG7hA/#scalars). I was able to get a 4% WER with CTC decoding.

I then tried training a pruned_transducer_stateless2 model (**with** speed perturbation), but training looks weird: [tensorboard](https://tensorboard.dev/experiment/vIDn5DidTga1MgMq20Ug4Q/#scalars). I am not sure if this is because of the 3x speed perturbation (I'm now training a model without speed perturbation to verify). I was hoping someone would be able to suggest what may be the reason for these periodic ups and downs.

Adding screenshot of training curve here:

image

(Please ignore the README files etc. for now since this is only a rough draft.)

### CODE REVIEW

**Comment by pzelasko** [comment]
Try pre-shuffling the input cutset like this: `gunzip -c train_cuts.jsonl.gz | shuf | gzip -c > train_cuts_shuf.jsonl.gz` and re-trying. `DynamicBucketingSampler` has a small-ish reservoir-sampling shuffle buffer of 10k cuts, and the data is likely sorted by recording sessions / speakers / topics / etc. which is why you see these patterns. Pre-shuffling will likely help.

**Comment by desh2608** [comment]
> gunzip -c train_cuts.jsonl.gz | shuf | gzip -c > train_cuts_shuf.jsonl.gz

Thanks for the suggestions. The pre-shuffling idea looks like the simplest, so I'll try that first and see how it goes.

**Comment by desh2608** [comment]
Pre-shuffling the training cuts seems to have fixed the issue. New training curve after 1 epoch (in blue):
image

**Comment by pzelasko** [comment]
Very cool. Can you please compare the WER of both runs, at different epochs and after whole training? I'm curious if it actually affects the performance of the system in a significant way or not.

**Comment by desh2608** [comment]
> Very cool. Can you please compare the WER of both runs, at different epochs and after whole training? I'm curious if it actually affects the performance of the system in a significant way or not.

For the original training (orange curve), the WER on a small dev set (decoded with `fast_beam_search`) after first 4 epochs are:

| Epoch | WER  |
|-------|------|
| 1     | 4.26 |
| 2     | 3.77 |
| 3     | 3.58 |
| 4     |  3.28  |

As a comparison, I got about ~4% WER with a conformer-CTC model (but trained without speed perturbation). The corrected model hasn't trained far enough yet so I haven't tried decoding with that one, but it seems the periodicity doesn't seem to have too much (if any) impact on the model performance.

**Comment by desh2608** [comment]
@csukuangfj I have been wondering recently about the batch size (`max-duration`) that we can use for the pruned transducer models. Even with 4 GPUs of 24G memory each, I am only able to use `--max-duration 60`, else I run into OOM errors which are thrown like [this](https://github.com/k2-fsa/icefall/issues/247). With 8 GPUs of 32G memory each, I run into memory error with `--max-duration 150`. However, I have seen in the LibriSpeech recipe that you use batch sizes of up to 300 when training with 8 GPUs?

**Comment by csukuangfj** [comment]
> @csukuangfj I have been wondering recently about the batch size (`max-duration`) that we can use for the pruned transducer models. Even with 4 GPUs of 24G memory each, I am only able to use `--max-duration 60`, else I run into OOM errors which are thrown like [this](https://github.com/k2-fsa/icefall/issues/247). With 8 GPUs of 32G memory each, I run into memory error even with `--max-duration 120`. However, I have seen in the LibriSpeech recipe that you use batch sizes of up to 300 when training with 8 GPUs?
> 
> 
> 
> (Related to @wgb14's observation [here](https://github.com/k2-fsa/icefall/pull/318#issuecomment-1103253489))

What's the distribution of your utterance duration. You may need to filter out short and long utterances.

**Comment by desh2608** [comment]
> What's the distribution of your utterance duration? You may need to filter out short and long utterances.

```python
In [4]: cuts = load_manifest_lazy('data/manifests/cuts_train.jsonl.gz')

In [5]: cuts.describe()
Cuts count: 5886320
Total duration (hours): 15070.1
Speech duration (hours): 15070.1 (100.0%)
***
Duration statistics (seconds):
mean    9.2
std     2.8
min     4.6
25%     6.9
50%     8.9
75%     11.2
99%     16.0
99.5%   16.3
99.9%   16.6
max     16.7
```
I would say it's pretty uniform.

**Comment by danpovey** [comment]
When you get an error, please print out the supervision object for the minibatch that fails.
Then we can figure out what the issue is specifically, e.g. is the batch made of short utts, or is it length mismatch,
or is it long utts?

**Comment by desh2608** [comment]
I finished training a pruned transducer model and here are the decoding results on the evaluation set:

| Decoding method           | val WER    | 
|---------------------------|------------|
| greedy search             | 2.40       |
| beam search               | 2.24       |
| modified beam search      | 2.30       |
| fast beam search          | 2.35       |

Here is the tensorboard: [link](https://tensorboard.dev/experiment/ExSoBmrPRx6liMTGLu0Tgw/#scalars)

As a comparison, the SPGISpeech paper reports 2.6% using NeMo's conformer CTC and 2.3% using ESPNet conformer encoder-decoder model. I believe at this point the recipe is ready to be merged.

@csukuangfj are there some instructions for how to prepare the model to be uploaded on HuggingFace? For instance, the above WERs are using `--avg-last-n=10`, so I suppose I need to average those models and prepare a "final" checkpoint that should be uploaded?

**Comment by csukuangfj** [comment]
> For instance, the above WERs are using --avg-last-n=10, so I suppose I need to average those models and prepare a "final" checkpoint that should be uploaded?

Please have a look at https://github.com/k2-fsa/icefall/blob/master/egs/librispeech/ASR/pruned_transducer_stateless2/export.py

You need to provide `--iter xx --avg 10`, where `xx` is your latest `checkpoint-xx.pt`.

After running `export.py`, you should get a file `pretrained.pt`, which can be uploaded to huggingface.

To upload files to huggingface, you need to first register an account on it. After creating an account, you can create a repo on it, clone it to your local computer, copy files to your local repo, and then use `git push` to upload them to huggingface.

One thing to note is that you have to run `sudo apt-get install git-lfs` before cloning the repo from huggingface. 
(`*.pt` files are tracked by git lfs by default in the cloned repo.)

**Comment by desh2608** [comment]
I have uploaded the pretrained model to HF. This PR is ready for review.

**Comment by csukuangfj** [comment]
@desh2608 

From https://huggingface.co/desh2608/icefall-asr-spgispeech-pruned-transducer-stateless2/blob/main/log/modified_beam_search/errs-dev-beam_size_4-epoch-28-avg-15-beam-4.txt
there are some insertions at the end of utterances.

You can fix the the insertion errors at end of utterances by using https://github.com/k2-fsa/icefall/pull/358
and re-run the decoding for greedy search and modified beam search. I think it helps reduce WERs.

You need to
(1) Use https://github.com/k2-fsa/icefall/blob/master/egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py
(2) In your decode.py, use

https://github.com/k2-fsa/icefall/blob/0f180b3ce20c3e704ef46972aa61816fd05a42ac/egs/librispeech/ASR/pruned_transducer_stateless2/decode.py#L270-L274

https://github.com/k2-fsa/icefall/blob/0f180b3ce20c3e704ef46972aa61816fd05a42ac/egs/librispeech/ASR/pruned_transducer_stateless2/decode.py#L278-L282

That is, add `encoder_out_lens=encoder_out_lens`.

File: egs/spgispeech/ASR/pruned_transducer_stateless2/decode.py [hunk 1]
```diff
@@ -0,0 +1,534 @@
...
+from collections import defaultdict
+from pathlib import Path
+from typing import Dict, List, Optional, Tuple
+
+import k2
+import sentencepiece as spm
+import torch
+import torch.nn as nn
+from asr_datamodule import SPGISpeechAsrDataModule
+from beam_search import (
```
Comment (csukuangfj) [inline egs/spgispeech/ASR/pruned_transducer_stateless2/decode.py:66]: If you are going to synchronize with the latest master, please use
```
vimdiff ./decode.py  /path/to/librispeech/ASR/pruned_transducer_stateless2/decode.py
```

to find the differences.

Comment (csukuangfj) [reply egs/spgispeech/ASR/pruned_transducer_stateless2/decode.py:66]: Please also fix the style issues reported by GitHub actions.

You can fix them locally by following
https://github.com/k2-fsa/icefall/blob/master/docs/source/contributing/code-style.rst

**Comment by desh2608** [comment]
> @desh2608
> 
> From https://huggingface.co/desh2608/icefall-asr-spgispeech-pruned-transducer-stateless2/blob/main/log/modified_beam_search/errs-dev-beam_size_4-epoch-28-avg-15-beam-4.txt there are some insertions at the end of utterances.
> 
> You can fix the the insertion errors at end of utterances by using #358 and re-run the decoding for greedy search and modified beam search. I think it helps reduce WERs.
> 
> You need to (1) Use https://github.com/k2-fsa/icefall/blob/master/egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py (2) In your decode.py, use
> 
> https://github.com/k2-fsa/icefall/blob/0f180b3ce20c3e704ef46972aa61816fd05a42ac/egs/librispeech/ASR/pruned_transducer_stateless2/decode.py#L270-L274
> 
> https://github.com/k2-fsa/icefall/blob/0f180b3ce20c3e704ef46972aa61816fd05a42ac/egs/librispeech/ASR/pruned_transducer_stateless2/decode.py#L278-L282
> 
> That is, add `encoder_out_lens=encoder_out_lens`.

The WER for modified beam search improved from 2.30% to 2.24% with this change. WER remained unchanged for greedy search.

File: egs/spgispeech/ASR/README.md [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ArXiv link: https://arxiv.org/abs/2104.02014
+
+## Performance Record
+
+| Decoding method           | val        |
+|---------------------------|------------|
+| greedy search             | 2.40       |
+| beam search               | 2.24       |
+| modified beam search      | 2.24       |
+| fast beam search          | 2.35       |
```
Comment (csukuangfj) [inline egs/spgispeech/ASR/README.md:31]: https://arxiv.org/pdf/2104.02014.pdf
The above paper also lists the CER. Shall we also do that?

Screen Shot 2022-05-14 at 21 59 13

Comment (desh2608) [reply egs/spgispeech/ASR/README.md:31]: Sure, I'll add those as well.

Comment (desh2608) [reply egs/spgispeech/ASR/README.md]: Done!

File: egs/spgispeech/ASR/pruned_transducer_stateless2/decode.py (hunk 1 above)
Comment (desh2608) [reply egs/spgispeech/ASR/pruned_transducer_stateless2/decode.py:70]: I think the remaining style issues are from librispeech code that I have soft-linked.

Comment (csukuangfj) [reply egs/spgispeech/ASR/pruned_transducer_stateless2/decode.py:70]: Ah, ok. Thanks! Merging.

**Comment by csukuangfj** [comment]
@desh2608 

Could you please upload a torchscript model to 
https://huggingface.co/desh2608/icefall-asr-spgispeech-pruned-transducer-stateless2/tree/main/exp

You can use `export.py --jit=1` to obtain a torchscript model.

I would like to add the pre-trained torchscript model to
https://huggingface.co/spaces/k2-fsa/automatic-speech-recognition

**Comment by desh2608** [comment]
> @desh2608 
> 
> 
> 
> Could you please upload a torchscript model to 
> 
> https://huggingface.co/desh2608/icefall-asr-spgispeech-pruned-transducer-stateless2/tree/main/exp
> 
> 
> 
> You can use `export.py --jit=1` to obtain a torchscript model.
> 
> 
> 
> I would like to add the pre-trained torchscript model to
> 
> https://huggingface.co/spaces/k2-fsa/automatic-speech-recognition

I will do it this weekend (I'm on an internship right now and have limited access to the JHU cluster).

**Comment by jtrmal** [comment]
@desh2608 let me know and I can add you to the k2-fsa hf repo
y.

**Comment by csukuangfj** [comment]
> > @desh2608 
> > 
> > 
> > 
> > Could you please upload a torchscript model to 
> > 
> > https://huggingface.co/desh2608/icefall-asr-spgispeech-pruned-transducer-stateless2/tree/main/exp
> > 
> > 
> > 
> > You can use `export.py --jit=1` to obtain a torchscript model.
> > 
> > 
> > 
> > I would like to add the pre-trained torchscript model to
> > 
> > https://huggingface.co/spaces/k2-fsa/automatic-speech-recognition
> 
> I will do it this weekend (I'm on an internship right now and have limited access to the JHU cluster).

Thanks!

**Comment by desh2608** [comment]
@csukuangfj sorry it took a while (I just got back from my internship). I have uploaded a jitted model [here](https://huggingface.co/desh2608/icefall-asr-spgispeech-pruned-transducer-stateless2/tree/main/exp).

**Comment by csukuangfj** [comment]
> @csukuangfj sorry it took a while (I just got back from my internship). I have uploaded a jitted model [here](https://huggingface.co/desh2608/icefall-asr-spgispeech-pruned-transducer-stateless2/tree/main/exp).

Thanks!
